crates/oxc_query/src/vertex.rs

```diff
@@ -0,0 +1,319 @@
+            Vertex::JSXElement(data) => data.ast_node.unwrap().id(),
+            Vertex::TypeAnnotation(data) => data.ast_node.unwrap().id(),
+            Vertex::VariableDeclaration(data) => data.ast_node.unwrap().id(),
+            Vertex::ReturnStatementAST(data) => data.ast_node.unwrap().id(),
+            Vertex::DefaultImport(_)
+            | Vertex::AssignmentType(_)
+            | Vertex::ClassMethod(_)
+            | Vertex::Expression(_)
+            | Vertex::File
+            | Vertex::InterfaceExtend(_)
+            | Vertex::JSXAttribute(_)
+            | Vertex::JSXExpressionContainer(_)
+            | Vertex::JSXFragment(_)
+            | Vertex::ObjectLiteral(_)
+            | Vertex::JSXText(_)
+            | Vertex::JSXSpreadChild(_)
+            | Vertex::JSXSpreadAttribute(_)
+            | Vertex::JSXOpeningElement(_)
+            | Vertex::PathPart(_)
+            | Vertex::URL(_)
+            | Vertex::Type(_)
+            | Vertex::SpecificImport(_)
+            | Vertex::Span(_)
+            | Vertex::SearchParameter(_)
+            | Vertex::ClassProperty(_) => {
+                unreachable!()
+            }
+        }
+    }
+
+    pub fn make_url(attr: &'a JSXAttribute<'a>) -> Option<Self> {
+        let Some(maybe_url) = jsx_attribute_to_constant_string(attr) else { return None };
+        let Ok(parsed_url) = Url::parse(&maybe_url) else { return None };
+        return Some(Vertex::URL(Rc::new(parsed_url)));
+    }
+
+    pub fn as_constant_string(&self) -> Option<String> {
+        match &self {
+            Vertex::Expression(expr) => expr_to_maybe_const_string(expr),
+            _ => None,
+        }
+    }
+}
+
+impl trustfall::provider::Typename for Vertex<'_> {
+    fn typename(&self) -> &'static str {
+        self.make_type_name()
+    }
+}
+
+impl<'a> From<AstNode<'a>> for Vertex<'a> {
+    fn from(ast_node: AstNode<'a>) -> Self {
+        match ast_node.kind() {
+            AstKind::ReturnStatement(return_statement) => Self::ReturnStatementAST(
+                ReturnStatementVertex { ast_node: Some(ast_node), return_statement }.into(),
+            ),
+            AstKind::JSXElement(element) => {
+                Self::JSXElement(JSXElementVertex { ast_node: Some(ast_node), element }.into())
+            }
+            AstKind::TSInterfaceDeclaration(interface) => {
+                Self::Interface(InterfaceVertex { ast_node: Some(ast_node), interface }.into())
+            }
+            AstKind::TSTypeAnnotation(type_annotation) => Self::TypeAnnotation(
+                TypeAnnotationVertex { ast_node: Some(ast_node), type_annotation }.into(),
+            ),
+            AstKind::VariableDeclarator(variable_declaration) => Self::VariableDeclaration(
+                VariableDeclarationVertex { ast_node: Some(ast_node), variable_declaration }.into(),
+            ),
+            AstKind::ModuleDeclaration(md)
+                if matches!(md, ModuleDeclaration::ImportDeclaration(_)) =>
```

Comment:
here's a little trick you might not have seen before:
```suggestion
            AstKind::ModuleDeclaration(md @ ModuleDeclaration::ImportDeclaration(_))
```
https://doc.rust-lang.org/book/ch18-03-pattern-syntax.html#-bindings

Reply:
can just skip that with inner pattern matching as done in [323dd48](https://github.com/web-infra-dev/oxc/pull/599/commits/323dd48adb6e1d89b10f84d86f6317876f92f367)

Reply:
Yeah for sure. I was giving you another option in case you needed the intermediate match as-is in addition to matching inside it.